### PR TITLE
feat(server): introduce rss oom limit

### LINKED ISSUE #3616: add additional guard like oom_deny_ratio but for rss
1. And gradually deprecate oom_deny_ratio
2. Also stop adding connections on the main port once we are low on memory (critical state).

### PR DESCRIPTION
**Pull request by adiholden**
fixes #3616

### CODE REVIEW

File: src/server/main_service.cc
Comment (adiholden) [inline src/server/main_service.cc:106]: @romange do you think this should be enabled by default or disabled by default?

File: src/server/main_service.cc
Comment (chakaz) [inline src/server/main_service.cc:109]:
```suggestion
          "When the ratio between maxmemory and RSS memory exceeds this value, commands marked as DENYOOM will fail with OOM error and new connections to non-admin port will be rejected. Negative value disables this feature.");
```

File: src/server/main_service.cc
Comment (chakaz) [inline src/server/main_service.cc:1073]: I'd wrap this logic, possibly including the `DLOG` part, in a function

File: src/server/main_service.cc
Comment (chakaz) [inline src/server/main_service.cc:1067]: Are we ok with reading 2 flags in the hot path?
We could have thread-local caches, and update them upon flag update

Comment (adiholden) [reply src/server/main_service.cc]: ok using thread local now

**Review by chakaz — APPROVED** [review]
Please wait for Roman's approval on the default value

File: src/server/main_service.cc
Comment (romange) [reply src/server/main_service.cc:106]: enabled but maybe increase the default ratio to `1.25` because it affects `maxmemory` semantics

Comment (romange) [reply src/server/main_service.cc:106]: we also need to provide explanation in release notes

File: src/server/main_service.cc
Comment (romange) [inline src/server/main_service.cc:1084]: it will be simpler if this function returns `bool`

File: src/server/main_service.cc
Comment (romange) [inline src/server/main_service.cc:959]: Please note that `EngineShard::RunPeriodic` runs shard_handler once in 100ms, even though the loop inside RunPeriodic is once in 1ms. 
So with this code now `server_family_.UpdateMemoryGlobalStats();` will run once in 100ms as well. I think it's not ok. I suggest that you move the 100ms check here and make it special only for `BreakStalledFlowsInShard()`  so that shard_handler will be called every iteration inside RunPeriodic.

File: src/server/main_service.cc
Comment (adiholden) [reply src/server/main_service.cc:106]: what if one upgrades to new version and does not use admin port. if he hits the rss limit he will not be able to even connect to the server as we will reject new connections

File: src/server/server_family.cc
Comment (romange) [inline src/server/server_family.cc:985]: I see you changed semantics of the previous code by updating `used_mem_peak` once in a second. I am confused now whether we need to run `UpdateMemoryGlobalStats` as frequently as 1ms. Maybe you are right and it's not needed.  But I think it's an opportunity to simplify this logic so that timing rules will be clearer. 

Lets run UpdateMemoryGlobalStats once in 100ms, and the whole shard_handler will run once in 100ms. But then lets remove the 1s restriction here and update `used_mem_peak`, `rss_mem_current`, `rss_mem_peak` once in 100ms as wll

File: src/server/server_family.cc
Comment (romange) [inline src/server/server_family.cc:1002]: you added `rss_oom_deny_ratio` to server state but here you use the flag.

File: src/server/main_service.cc
Comment (romange) [reply src/server/main_service.cc:106]: hmm, you are right but how we introduce this behavior then? I do think that some sane multiplier is needed. This can be eve 2,3 but if RSS jumps by a huge factor - it's not ok.

File: src/server/server_family.cc
Comment (adiholden) [reply src/server/server_family.cc:1002]: will fix

File: src/server/server_family.cc
Comment (adiholden) [reply src/server/server_family.cc]: ok I removed the logic to restrict this logic to run only once a second

File: src/server/server_family.h
Comment (romange) [inline src/server/server_family.h:340]: no need for `global_stats_update_time_`

**Review by romange — COMMENTED** [review]
lgtm

File: src/server/main_service.cc
Comment (adiholden) [reply src/server/main_service.cc]: maybe we should reject new connections only if admin port is set?

File: src/server/server_family.cc
Comment (romange) [inline src/server/server_family.cc:1000]: I know it's what we wanted by imagine the scenario where a community-run server hits this condition. In that case, most likely they won't be able to manage the server, nor clear the data or run INFO because they do not use the admin_port. 

I suggest to restrict this condition and disable the main listener only if there are multiple listeners, or even better - there is a priveleged listener

Comment (adiholden) [reply src/server/server_family.cc:1000]: https://github.com/dragonflydb/dragonfly/pull/3702#discussion_r1767375198

File: src/server/main_service.cc
Comment (romange) [reply src/server/main_service.cc]: sounds good

**Review by romange — APPROVED** [review]

**Pull request merged**
